File: contexts/typescript/services/code_host.ts
// Code host — opens pull requests. This fake records each PR under `.prs/<slug>/`
// (a human-readable patch + the patched file contents, so a reviewer can merge it).
// Swap for your git host's PR API (ESSENCE §4H).
import { existsSync, mkdirSync, readdirSync, statSync, writeFileSync } from "node:fs";
import path from "node:path";

export class LocalCodeHost {
  prsDir: string;
  constructor(prsDir: string) {
    this.prsDir = prsDir;
  }

  openPr(slug: string, title: string, body: string, diff: string, files: Record<string, string>): string {
    const d = path.join(this.prsDir, slug);
    mkdirSync(d, { recursive: true });
    writeFileSync(path.join(d, "pr.md"), `# ${title}\n\n${body}\n\n\`\`\`diff\n${diff}\`\`\`\n`);
    writeFileSync(path.join(d, "files.json"), JSON.stringify(files));
    return `file://${d}`;
  }

  listPrs(): string[] {
    if (!existsSync(this.prsDir)) return [];
    return readdirSync(this.prsDir)
      .filter((n) => statSync(path.join(this.prsDir, n)).isDirectory())
      .sort();
  }
}

File: contexts/python/services/code_host.py
"""Code host — opens pull requests. This fake records each PR under `.prs/<slug>/`
(a human-readable patch + the patched file contents, so a reviewer can merge it).
Swap for your git host's PR API (ESSENCE §4H)."""
import json
import os


class LocalCodeHost:
    def __init__(self, prs_dir: str) -> None:
        self.prs_dir = prs_dir

    def open_pr(self, slug: str, title: str, body: str, diff: str, files: dict) -> str:
        d = os.path.join(self.prs_dir, slug)
        os.makedirs(d, exist_ok=True)
        with open(os.path.join(d, "pr.md"), "w", encoding="utf-8") as f:
            f.write(f"# {title}\n\n{body}\n\n```diff\n{diff}```\n")
        with open(os.path.join(d, "files.json"), "w", encoding="utf-8") as f:
            json.dump(files, f)
        return f"file://{d}"

    def list_prs(self) -> list[str]:
        if not os.path.isdir(self.prs_dir):
            return []
        return sorted(n for n in os.listdir(self.prs_dir) if os.path.isdir(os.path.join(self.prs_dir, n)))
